### PR TITLE
Content directory

### PR DESCRIPTION
**Pull request by jfinkhaeuser**
Based off another branch, so makes sense only after #25 is merged.

However, this should be enough for a first stab at a content directory.

### CODE REVIEW

**Comment by jfinkhaeuser** [comment]
Tests are missing so far.

File: src/storage/content_directory.rs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+pub trait Trait: timestamp::Trait + system::Trait + DOTRTrait + MaybeDebug {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    type MetadataId: Parameter + Member + SimpleArithmetic + Codec + Default + Copy
+        + As<usize> + As<u64> + MaybeSerializeDebug + PartialEq;
+
+    // Schema ID should be defined in a different Trait in future
+    type SchemaId: Parameter + Member + SimpleArithmetic + Codec + Default + Copy
+        + As<usize> + As<u64> + MaybeSerializeDebug + PartialEq;
+
```
Comment (siman) [inline src/storage/content_directory.rs]: redundant extra empty line

File: src/storage/content_directory.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+                return Err(MSG_CREATOR_MUST_BE_MEMBER);
+            }
+
+            // TODO in future, we want the schema IDs to correspond to a schema
+            // registry, and validate the data. For now, we just allow the
+            // following schema IDs:
+            // 1 - Video (schema TBD, i.e. handled in UI only)
+            // 2 - Podcast (schema TBD, i.e. handled in UI only)
+            // Pseudocode
+            // if schema in (1, 2) {
```
Comment (siman) [inline src/storage/content_directory.rs]:
```suggestion
            // if schema not in (1, 2) {
```

File: src/storage/content_directory.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            // Ensure signed account
+            let who = ensure_signed(origin).clone().unwrap();
+
+            // Try t find metadata
+            let found = Self::metadata(id);
+            ensure!(!found.is_some(), MSG_METADATA_NOT_FOUND);
+
+            // Ensure it's the metadata creator who publishes
+            let mut data = found.unwrap();
+            if data.origin != who {
```
Comment (siman) [inline src/storage/content_directory.rs]:
```suggestion
            ensure!(data.origin == who, MSG_ONLY_OWNER_MAY_PUBLISH);
```

File: src/storage/content_directory.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+        pub fn add_metadata(origin, schema: T::SchemaId, metadata: Vec<u8>)
+        {
+            // Origin has to be a member
+            let who = ensure_signed(origin).clone().unwrap();
+            if !T::IsActiveMember::is_active_member(&who) {
+                return Err(MSG_CREATOR_MUST_BE_MEMBER);
+            }
+
+            // TODO in future, we want the schema IDs to correspond to a schema
```
Comment (siman) [inline src/storage/content_directory.rs]: Makes sense to mention a corresponding issue here in this TODO-comment:
https://github.com/Joystream/substrate-runtime-joystream/issues/20

File: src/storage/content_directory.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+decl_storage! {
+    trait Store for Module<T: Trait> as ContentDirectory {
+        // Start at this value
+        pub FirstMetadataId get(first_metadata_id) config(first_metadata_id): T::MetadataId = T::MetadataId::sa(DEFAULT_FIRST_METADATA_ID);
+
+        // Increment
+        pub NextMetadataId get(next_metadata_id) build(|config: &GenesisConfig<T>| config.first_metadata_id): T::MetadataId = T::MetadataId::sa(DEFAULT_FIRST_METADATA_ID);
+
+        // Mapping of Data object types
+        pub MetadataMap get(metadata): map T::MetadataId => Option<Metadata<T>>;
```
[the remark this answers is not in the document]
Comment (jfinkhaeuser) [reply src/storage/content_directory.rs]: Data is plural, technically... We could do `Metadata`, sure

File: src/storage/data_directory.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        fn deposit_event<T>() = default;
+
+        pub fn add_content(origin, data_object_type_id: <T as DOTRTrait>::DataObjectTypeId,
+                           id: T::ContentId, size: u64) {
+            // Origin has to be a member
+            let who = ensure_signed(origin).clone().unwrap();
```
[the remark this answers is not in the document]
Comment (jfinkhaeuser) [reply src/storage/data_directory.rs]: Will get to this, too

File: src/storage/content_directory.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (siman) [reply src/storage/content_directory.rs]: I mean `Map` is cool, but such as Rust is a typed language, it's quite clear that this item is a Map (until IDE works well, haha)

**Pull request merged**